File: 81-search-in-rotated-sorted-array-ii/81-search-in-rotated-sorted-array-ii.py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        left, right = 0, len(nums) - 1

        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                return True
            
            # case duplicate number
            if nums[left] == nums[mid] and nums[left] == nums[right]:
                left += 1
                right -= 1
                
            # left sorted
            elif nums[left] <= nums[mid]:
                if nums[left] <= target and nums[mid] > target:
                    right = mid - 1
                else:
                    left = mid + 1
                    
            # right sorted
            else:
                if nums[mid] < target and nums[right] >= target:
                    left = mid + 1
                else:
                    right = mid - 1
                
        return False
                    
        

File: 81-search-in-rotated-sorted-array-ii/81-search-in-rotated-sorted-array-ii.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Boolean}
def search(nums, target)
    left, right = 0, nums.length - 1
    
    while left <= right do
        mid = (left + right) / 2
        
        return true if nums[mid] == target
        
        if nums[left] == nums[mid] && nums[left] == nums[right]
            left += 1
            right -= 1
        elsif nums[left] <= nums[mid]
            if nums[left] <= target && nums[mid] > target
                right = mid - 1
            else
                left = mid + 1
            end
        else
            if nums[right] >= target  && nums[mid] < target
                left = mid + 1
            else
                right = mid - 1
            end
        end
    end
    
    false
end